win,fs: retry if uv_fs_rename fails

bzoz opened the pull request:
On Windows rename operation can fail randomly in presence of antivirus or indexing software. Popular workaround is to retry the operation after small delay, like in [graceful-fs](https://github.com/isaacs/node-graceful-fs/blob/master/polyfills.js#L88-L116), it also appear in other places, like [here](https://github.com/davidmarkclements/0x/pull/128/files) and [here](https://stackoverflow.com/questions/964920/is-it-possible-to-reasonably-workaround-an-antivirus-scanning-the-working-direct).

This PR implements this workaround in libuv. It makes `uv_fs_rename` retry up to four times with 250ms delay between attempts before giving up.

bzoz commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1023/

bnoordhuis commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,32 @@ static void fs__fstat(uv_fs_t* req) {
+
+  tries = 0;
+
+rename_retry:
```
Why don't you use `for (tries = 0; tries < 3; tries += 1)` here?

mcollina commented:
Either this should live in libuv, or it should live in Node.js. I think Node.js `fs.rename()` operation should work on Windows. Without this mechanism, it is completely unreliable and mostly unusable. It is also very hard to debug if you are not familiar with the problem.

Doing it on Node.js have the added benefit of scheduling the retry over the event loop, freeing the thread for other stuff.

bzoz commented:
IMHO this should be added to libuv. Other projects will encounter this issue sooner or later. Fixing this sounds like something libuv should do.

@cjihrig: sleep time and retries are arbitrary

bzoz commented:
Ping? This would solve a real-world Windows issue.

@libuv/collaborators

vtjnash commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,32 @@ static void fs__fstat(uv_fs_t* req) {
-  SET_REQ_RESULT(req, 0);
+  sys_errno = GetLastError();
+  result = uv_translate_sys_error(sys_errno);
+  if ((result == UV_EBUSY || result == UV_EPERM) &&
```
The node-graceful-fs code you linked to appears to check for a different error. Is that intentional? 
```
            && (er.code === "EACCES" || er.code === "EPERM")
```

It also appears that it only retries as long as `stat(to)` returns `ENOENT`, fwiw. This was added in https://github.com/isaacs/node-graceful-fs/issues/98 (wherein the delay was also changed from 1 s to 60 s).

vtjnash commented:
To answer part of my own question (and in response to JuliaLang/julia#29658 reporting a similar sounding issue), I wrote a quick test to simulate the failure situation described here to investigate how Julia's implementation would be affected by this PR. As expected, the copy fails with EBUSY, the mv succeeds, and then the delete fails (again with EBUSY). This doesn't happen with libuv or nodejs, since it opts into more posix-compliant semantics (https://github.com/nodejs/node-v0.x-archive/issues/1449), unless specifically requested (https://github.com/libuv/libuv/issues/1605, https://github.com/libuv/libuv/commit/1c4de1916e36f8462c48a36ce7c88b247465f3cf), while Julia is currently still using the CRT directly to implement `fs.open`.

bzoz commented:
Updated with some changes, PTAL

In my testings, the rename would fail with `UV_EBUSY` when the source file is opened or `UV_EPERM` if the target is in use. I've now also added `UV_EACCES` since `graceful-fs` and others also test for it.

As for the target being a directory - added a check for that, and if both source and target are directories, we will now try to delete target (if it is empty) to make everything work. While at it, I've also added some checks (like for read-only target) to make rename fail early.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1071/
CI-node: https://ci.nodejs.org/view/libuv/job/libuv-in-node/68/

thefourtheye commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
+      case 0:
+      case UV_ENOENT:
+        /* Folder removed or did not exist at all. */
+        try_rmdir = 0;
```
When copying directories, mv might have copied a few files and subdirectories before failing. In that case, shouldn't we try to delete the destination again?

bzoz replied:
The directory is renamed, it happens at once.

santigimeno commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
+  int sys_errno;
+  int result;
+  int try_rmdir;
+  WCHAR *src, *dst;
```
Style: align *  to the left

santigimeno commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
+  }
+  req->sys_errno_ = sys_errno;
+  req->result = result;
+  return;
```
Do we need a return here?

santigimeno commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
+  for (tries = 0; tries < UV__RENAME_RETRIES; ++tries) {
+    if (tries > 0) {
+      Sleep(UV__RENAME_WAIT);
+    }
```
Remove braces here?

santigimeno commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
+        result = 0;
+      } else {
+        result = uv_translate_sys_error(_doserrno);
+      }
```
And here?

santigimeno commented on src/win/fs.c:
```diff
@@ -1329,12 +1331,85 @@ static void fs__fstat(uv_fs_t* req) {
+    result = uv_translate_sys_error(sys_errno);
+    if (result != UV_EBUSY && result != UV_EPERM && result != UV_EACCES) {
+      break;
+    }
```
And here

bzoz commented:
Updated, PTAL

bzoz commented:
Resumed because aix (which looked unrelated): https://ci.nodejs.org/view/libuv/job/libuv-test-commit/1083/

bzoz commented:
CI is green

bzoz commented:
Landed in https://github.com/libuv/libuv/commit/e94c184c7c4a18f3de569c97caeb83f4ff98a4b2
